libopae/plugins/xfpga/feature_plugins/feature_api.c

```diff
@@ -0,0 +1,390 @@
+			if ((!uuid_is_null(prop->guid) && (uuid_compare(prop->guid, guid) == 0))
+				|| (uuid_is_null(prop->guid))) { 
+				if (*num_matches < max_tokens) {
+					fpga_feature_token tmp = 0;
+					feature_adapter_table * adapter;
+					
+					result = xfpga_fpgaCloneFeatureToken((fpga_feature_token)_ftoken, &tmp);
+					if (result	!= FPGA_OK) {
+						FPGA_MSG("Error cloning token");
+					}
+					adapter = get_feature_plugin_adapter(guid);
+					wrapped_feature_token *wt = 
+					     allocate_wrapped_feature_token(tmp, adapter);
+					tokens[*num_matches] = wt;
+				}
+				++(*num_matches);
+			}
+		}
+
+		// Move to the next feature header 
+		offset = offset + dfh.next_header_offset;
+	} while (!dfh.eol);
+
+	return result;
+}
+
+fpga_result __FPGA_API__ xfpga_fpgaDestroyFeatureToken(fpga_feature_token *token)
+{
+	wrapped_feature_token *wrapped_token;
+
+	ASSERT_NOT_NULL(token);
+
+	wrapped_token = validate_wrapped_feature_token(*token);
+
+	ASSERT_NOT_NULL(wrapped_token);
+	
+	free(wrapped_token->feature_token);
+	wrapped_token->feature_token = NULL;
+
+	destroy_wrapped_feature_token(wrapped_token);
+
+	return FPGA_OK;
+	
+}
+
+fpga_result __FPGA_API__ xfpga_fpgaFeaturePropertiesGet(fpga_feature_token token,
+	fpga_feature_properties *prop)
+{
+	fpga_result res = FPGA_OK;
+	wrapped_feature_token *wrapped_token;
+	struct _fpga_feature_token *_ftoken;
+	errno_t e;
+
+	wrapped_token = validate_wrapped_feature_token(token);
+	_ftoken = (struct _fpga_feature_token *)wrapped_token->feature_token;
+	
+	ASSERT_NOT_NULL(wrapped_token);
+	ASSERT_NOT_NULL(prop);
+	ASSERT_NOT_NULL(_ftoken);
+
+	if (_ftoken->magic != FPGA_FEATURE_TOKEN_MAGIC) {
+		FPGA_ERR("Invalid feature token");
+		return FPGA_INVALID_PARAM;
+	}
+
+	prop->type = (fpga_feature_type )_ftoken->feature_type;
+	e = memcpy_s(prop->guid, sizeof(fpga_guid), _ftoken->feature_guid,
+		sizeof(fpga_guid));
+
+	if (EOK != e) {
+		FPGA_ERR("memcpy_s failed");
+	}
+
+	return res;
+}
+
+fpga_result __FPGA_API__ xfpga_fpgaFeatureOpen(fpga_feature_token token, int flags,
+                            void *priv_config, fpga_feature_handle *handle)
+{
+	fpga_result res;
+	fpga_result cres = FPGA_OK;
+	wrapped_feature_token *wrapped_token;
+	fpga_feature_handle feature_handle = NULL;
+	wrapped_feature_handle *wrapped_handle;
+
+	wrapped_token = validate_wrapped_feature_token(token);
+
+	ASSERT_NOT_NULL(wrapped_token);
+	ASSERT_NOT_NULL(handle);
+	ASSERT_NOT_NULL_RESULT(wrapped_token->adapter_table->fpgaFeatureOpen,
+			       FPGA_NOT_SUPPORTED);
+
+	res = wrapped_token->adapter_table->fpgaFeatureOpen(wrapped_token->feature_token,
+						    flags, priv_config, &feature_handle);
+
+	ASSERT_RESULT(res);
+
+	wrapped_handle = allocate_wrapped_feature_handle(
+		wrapped_token, feature_handle, wrapped_token->adapter_table);
+
+	if (!wrapped_handle) {
+		FPGA_ERR("malloc failed");
+		res = FPGA_NO_MEMORY;
+		cres = wrapped_token->adapter_table->fpgaFeatureClose(feature_handle);
```

Comment:
You haven't verified that fpgaFeatureClose is non-NULL before calling it.

Reply:
Will fix